### PR TITLE
Version 4 codemods

### LINKED ISSUE #2983: code mod for breaking syntax changes
<details>
  <summary>useQuery, useInfiniteQuery, queryClient.fetchQuery, queryClient.prefetchQuery, queryClient.fetchInfiniteQuery, queryClient.prefetchInfiniteQuery, useInfiniteQuery</summary>

```
input:
  useQuery('todos', fetchTodos)
output:
  useQuery(['todos'], fetchTodos)

input:
  useQuery(['todos'], fetchTodos, { staleTime: 1000 })
output:
  useQuery(['todos'], fetchTodos, { staleTime: 1000 })

input:
  useQuery(['todos'], { queryFn: fetchTodos, staleTime: 1000 })
output:
  useQuery(['todos'], { queryFn: fetchTodos, staleTime: 1000 })

input:
  useQuery('todos', { queryKey: ['notTodos'], queryFn: fetchTodos, staleTime: 1000 })
output:
  useQuery(['todos'], { queryKey: ['notTodos'], queryFn: fetchTodos, staleTime: 1000 })

input:
  useQuery({ queryKey: 'todos', staleTime: 1000 })
output:
  useQuery({ queryKey: ['todos'], staleTime: 1000 })
  
input:
  const key = 'todos'
  useQuery(key, { queryFn: fetchTodos, staleTime: 1000 })
output:
  useQuery({ queryKey: [key], queryFn: fetchTodos, staleTime: 1000 })

input:
  const queryKey = 'todos'
  useQuery({ queryKey, queryFn: fetchTodos, staleTime: 1000 })
output:
  useQuery({ queryKey: [queryKey], queryFn: fetchTodos, staleTime: 1000 })
  
input:
  const key = ['todos']
  useQuery(key, { queryFn: fetchTodos, staleTime: 1000 })
output:
  useQuery(key, { queryFn: fetchTodos, staleTime: 1000 })

input:
  const queryKey = ['todos']
  useQuery({ queryKey, queryFn: fetchTodos, staleTime: 1000 })
output:
  useQuery({ queryKey, queryFn: fetchTodos, staleTime: 1000 })

input:
  const queryKey = createKey()
  useQuery({ queryKey, queryFn: fetchTodos, staleTime: 1000 })
output:
  useQuery({ queryKey, queryFn: fetchTodos, staleTime: 1000 })

input:
  import { useQuery as useReactQuery } from 'react-query'
  const useQuery = (key, fn, options) => useReactQuery(key, fn, { ...options, staleTime: 1000 })
  useQuery('todos', fetchTodos, {})
output:
  import { useQuery as useReactQuery } from 'react-query'
  const useQuery = (key, fn, options) => useReactQuery(key, fn, { ...options, staleTime: 1000 })
  useQuery(['todos'], fetchTodos, {})

input:
  import * as RQ from 'react-query'
  const useQuery = (key, fn, options) => RQ.useQuery(key, fn, { ...options, staleTime: 1000 })
  useQuery('todos', fetchTodos, {})
output:
  import * as RQ from 'react-query'
  const useQuery = (key, fn, options) => RQ.useQuery(key, fn, { ...options, staleTime: 1000 })
  useQuery(['todos'], fetchTodos, {})
```
</details>

<details>
  <summary>
    queryClient.invalidateQueries, queryClient.refetchQueries, queryClient.resetQueries, queryClient.removeQueries, queryClient.cancelQueries
  </summary>

```
input:
  const queryClient = useQueryClient()
  queryClient.invalidateQueries('todos')
output:
  const queryClient = useQueryClient()
  queryClient.invalidateQueries(['todos'])

input:
  const queryClient = useQueryClient()
  queryClient.invalidateQueries('todos', { exact: true })
output:
  const queryClient = useQueryClient()
  queryClient.invalidateQueries(['todos'], { exact: true })

input:
  const queryClient = useQueryClient()
  queryClient.invalidateQueries(['todos'], { exact: true }, options)
output:
  const queryClient = useQueryClient()
  queryClient.invalidateQueries(['todos'], { exact: true }, options)

input:
  useQueryClient().invalidateQueries({ queryKey: 'todos', exact: true })
output:
  useQueryClient().invalidateQueries({ queryKey: ['todos'], exact: true })

input:
  useQueryClient().invalidateQueries({ queryKey: ['todos'], exact: true })
output:
  useQueryClient().invalidateQueries({ queryKey: ['todos'], exact: true })

input:
  const queryKey = 'todos'
  useQueryClient().invalidateQueries({ queryKey, exact: true })
output:
  useQueryClient().invalidateQueries({ queryKey: [queryKey], exact: true })

input:
  const queryKey = ['todos']
  useQueryClient().invalidateQueries({ queryKey, exact: true })
output:
  useQueryClient().invalidateQueries({ queryKey, exact: true })
```
</details>

<details>
  <summary>queryClient.getQueryData, queryClient.getQueryState, queryCache.find, queryClient.getQueriesData, queryClient.isFetching, useIsFetching, useIsMutating, queryCache.findAll, queryClient.setQueriesData</summary>

```
input:
  const queryClient = useQueryClient()
  queryClient.getQueryData('key', { exact: true })
output:
  const queryClient = useQueryClient()
  queryClient.getQueryData(['key'], { exact: true })

input:
  const queryClient = useQueryClient()
  queryClient.getQueryData(['key'], { exact: true })
output:
  const queryClient = useQueryClient()
  queryClient.getQueryData(['key'], { exact: true })
```
</details>

<details>
  <summary>queryClient.setQueryData, queryClient.setMutationDefaults, queryClient.getMutationDefaults, queryClient.setQueryDefaults, queryClient.getQueryDefaults</summary>

```diff
- fn(key, updater, options)
+ fn([key], updater, options)
```
</details>

<details>
  <summary>useMutation</summary>

same as `useQuery`, except that `queryKey` is `mutationKey`

</details>

<details>
  <summary>useQueries</summary>

An easy one because useQueries only gets an array passed, which should become an object. Nothing within query1 / query2 / query3 needs to change.

```diff
- useQueries([query1, query2, query3])
+ useQueries({ queries: [query1, query2, query3])
```

after that, inside the `queries` object, we need to change `queryKey` to be an array, like for `useQuery` with the object syntax.
</details>

### PR DESCRIPTION
**Pull request by balazsmatepetro**
closes #2983

### CODE REVIEW

**Comment by balazsmatepetro** [comment]
@TkDodo please note this draft still doesn't support template literals. I can add support for template literals as well if you wish. :)

**Review by TkDodo — COMMENTED** [review]
I don't see the codemod tests being picked up by jest when doing `npm test` ...

File: package.json [hunk 1]
```diff
@@ -67,6 +67,7 @@
   "dependencies": {
     "@babel/runtime": "^7.5.5",
     "broadcast-channel": "^3.4.1",
+    "jscodeshift": "^0.13.1",
```
Comment (TkDodo) [inline package.json:70]: I don't think we should add this as a dependency? If the plan is to run the codemods via `npx jscodeshift`, it also shouldn't be necessary I believe?

Comment (balazsmatepetro) [reply package.json]: That was a mistake, I wanted to add it to the `devDependencies`. 😅 

980f30f9957a2c3681489d1a64e068f9861b8cde

**Comment by balazsmatepetro** [comment]
> I don't see the codemod tests being picked up by jest when doing `npm test` ...

It should be OK now.

80adc1b7ed7393cd982a71c9c79367dd59494e81

**Comment by balazsmatepetro** [comment]
@TkDodo could you take a look at the test cases, please? I think all scenarios are covered, but maybe I'm mistaken.

File: package.json (hunk 1 above)
Comment (TkDodo) [reply package.json]: the linked commit only shows changes to `package.json` - is the `yarn.lock` file up to date? I can't see it, but it shouldn't have `jscodeshift` as dependency, but only as dev-dependency, right?

**Comment by TkDodo** [comment]
@balazsmatepetro I think we should also run the codemod on the examples, because we need them to be updated as well. We can do this in this PR I'd say.

Also, please add documentation on how to apply this codemod to migration guide

**Review by TkDodo — COMMENTED** [review]
in the PR pipeline, I can see lots of log warnings:

https://github.com/tannerlinsley/react-query/runs/5082880146?check_suite_focus=true

like:

```
In file /home/runner/work/react-query/react-query/codemods/v4/__testfixtures__/parameter-is-identifier.input.tsx at line 26 the type of the `queryKey` couldn't be recognized, so the codemod couldn't be applied. Please migrate manually.


      148 |     } catch (error) {
      149 |       if (error.name === 'UnprocessableKeyError') {
    > 150 |         console.warn(error.message)
          |                 ^
      151 |         return node
      152 |       }
      153 | 

      at NodePath.<anonymous> (v4/utils/replacers/key-replacer.js:150:17)
      at NodePath.<anonymous> (../node_modules/jscodeshift/dist/collections/Node.js:142:47)
      at ../node_modules/jscodeshift/dist/Collection.js:75:36
          at Array.forEach (<anonymous>)
      at Collection.forEach (../node_modules/jscodeshift/dist/Collection.js:74:18)
      at Collection.replaceWith (../node_modules/jscodeshift/dist/collections/Node.js:140:17)
      at Collection.typedMethod (../node_modules/jscodeshift/dist/Collection.js:[41](https://github.com/tannerlinsley/react-query/runs/5082880146?check_suite_focus=true#step:5:41)3:[43](https://github.com/tannerlinsley/react-query/runs/5082880146?check_suite_focus=true#step:5:43))
```

I think this is rather distractive when looking at tests. I'm guessing this is expected, but maybe we can suppress the warnings when node_env is test?

File: docs/src/pages/guides/migrating-to-react-query-4.md [hunk 2]
```diff
@@ -18,6 +18,20 @@ To streamline all apis, we've decided to make all keys Arrays only:
 + useQuery(['todos'], fetchTodos)
 ```
 
+To make this migration easier we decided to deliver a codemod. You can easily apply it by using one (or both) of the following commands.
```
Comment (TkDodo) [inline docs/src/pages/guides/migrating-to-react-query-4.md:21]:
```suggestion

#### Codemod

To make this migration easier, we decided to deliver a codemod.

> The codemod is a best efforts attempt to help you migrate the breaking change. Please review the generated code thoroughly! Also, there are edge cases that cannot be found by the code mod, so please keep an eye on the log output.

You can easily apply it by using one (or both) of the following commands:
```

**Comment by balazsmatepetro** [comment]
> in the PR pipeline, I can see lots of log warnings:

It should be okay now: c5192a4e855478cd1718b96d9a48eaf9d2f4cd9d

**Review by TkDodo — APPROVED** [review]

**Comment by balazsmatepetro** [comment]
> I think we should also run the codemod on the examples, because we need them to be updated as well. We can do this in this PR I'd say.

Done: 787f7d98a98a981e08f3fbf17c1ca538a50f4a99

> Also, please add documentation on how to apply this codemod to migration guide

Done: a672f6484cba8a55a1645a2f9ff13cbcac743175 and I accepted your suggestion as well. :)

**Comment by balazsmatepetro** [comment]
@all-contributors
please add @balazsmatepetro for code.

**Comment by TkDodo** [comment]
@all-contributors please add @balazsmatepetro for code.

**Comment by balazsmatepetro** [comment]
@all-contributors please add @balazsmatepetro for code.

**Comment by balazsmatepetro** [comment]
@all-contributors please add @balazsmatepetro for code.

**Pull request merged**